_posts/2020/2020-10-28-upgrading-shared-conda-installation.markdown

```diff
@@ -0,0 +1,167 @@
+---
+author: kjthomps
+title: "Upgrading the Shared Conda Installation at ISRDI"
+teaser: "A shared conda installation provides a multitude of benefits for an organization, but upgrading it can be a challenge. This is that journey."
+categories: Code
+tags:
+- Python
+- Conda
+- Reproducibility
+- PackageManagement
+- DevOps
+---
+ 
+## Background
+
+The Python ecosystem offers many ways to manage Python and third-party libraries, pip + virtualenv being a very common solution.
+At ISRDI, we have a fairly specific need not met by most of these solutions: we develop and deploy a large amount of software, mostly command-line tools, for our researchers to process large-scale demographic data.
+These scripts reside on multi-user linux servers, and it's a requirement that they perform consistently and reproducibly for all team members.
+For this purpose, we've arrived at conda as being the solution that works best for our needs, as it has the ability to create named Python environments in which dependencies can be managed by safe and controlled means. 
```

Comment:
* We want to deploy command-line tools that can trivially invoke python in a named environment. Have a look at any of the bash wrappers in production script emporium, e.g. `cat /pkg/ipums/programming/perl_ipums/freq_quick`
* We want the ability to build binary, installable, portable packages for internal libraries. Conda build offers a cleaner (barely!) solution than the pip/virtualenv/wheel world.
* The Jupyter ecosystem is tailored to nicely work with named conda environments, particularly for managing kernels shared amongst all users. It is possible to do this with virtualenv, but for this use case definitely not easier.
* Conda is language agnostic and can install non-python libraries. There are at least a few places where we've pulled in non-Python packages into an environment (c++, R)
* Lastly, we're already heavily invested in conda as our environment management tool. Changing to virtualenv+pip would be a pretty big effort. I'm not closed to it by any means, I kind of hate conda, but it does offer some things (see other bullet points) that pip+virtualenv either doesn't support or does clumsily.

Reply:
Fran is asking a good question and one we should probably explore again down the line. For the purposes of the blog post, it might be better to retract language that conda is the only viable solution for us, and instead say conda is the technology we chose as best suited for our needs.
I think I'd love to see a conda-less future. It's a royal pain to administer. I just think the alternatives are just a different toothache.